wtxmgr/tx_test.go

```diff
@@ -1364,3 +1364,205 @@ func TestRemoveUnminedTx(t *testing.T) {
 			len(unminedTxns))
 	}
 }
+
+// testInsertDoubleSpendTx is a helper test which double spends an output. The
+// boolean parameter indicates whether the first spending transaction should be
+// the one confirmed. This test ensures that when a double spend occurs and both
+// spending transactions are present in the mempool, if one of them confirms,
+// then the remaining conflicting transactions within the mempool should be
+// removed from the wallet's store.
+func testInsertMempoolDoubleSpendTx(t *testing.T, first bool) {
+	store, db, teardown, err := testStore()
+	defer teardown()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dbtx, err := db.BeginReadWriteTx()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer dbtx.Commit()
+
+	ns := dbtx.ReadWriteBucket(namespaceKey)
+
+	// We'll start off the test by creating a new coinbase output at height
+	// 100 and inserting it into the store.
+	b100 := BlockMeta{
+		Block: Block{Height: 100},
+		Time:  time.Now(),
+	}
+
+	cb := newCoinBase(1e8)
+	cbRec, err := NewTxRecordFromMsgTx(cb, b100.Time)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := store.InsertTx(ns, cbRec, &b100); err != nil {
+		t.Fatal(err)
+	}
+
+	// Then, we'll create two spends from the same coinbase output, in order
+	// to replicate a double spend scenario.
+	firstSpend := spendOutput(&cbRec.Hash, 0, 5e7, 5e7)
+	firstSpendRec, err := NewTxRecordFromMsgTx(firstSpend, time.Now())
+	if err != nil {
+		t.Fatal(err)
+	}
+	secondSpend := spendOutput(&cbRec.Hash, 0, 4e7, 6e7)
+	secondSpendRec, err := NewTxRecordFromMsgTx(secondSpend, time.Now())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// We'll insert both of them into the store without confirming them.
+	if err := store.InsertTx(ns, firstSpendRec, nil); err != nil {
+		t.Fatal(err)
+	}
+	if err := store.InsertTx(ns, secondSpendRec, nil); err != nil {
+		t.Fatal(err)
+	}
+
+	// Ensure that both spends are found within the unconfirmed transactions
+	// in the wallet's store.
+	unminedTxs, err := store.UnminedTxs(ns)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(unminedTxs) != 2 {
+		t.Fatalf("expected 2 unmined txs, got %v", len(unminedTxs))
+	}
+
+	// Then, we'll confirm the second spend with a height deep enough that
```

Comment:
comment should say either first or second based on the boolean.

Reply:
Fixed.